oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java

```diff
@@ -0,0 +1,175 @@
+     *                          <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">
+     *                          HTTP Authentication Scheme Registry</a>.
+     */
+    protected AbstractOAuth2AuthorizationGrantBuilder clientAuthorization(
+            Supplier<String> authorizationSupplier, String authorizationType) {
+        clientAuthorization = ClientAuthorization.ofAuthorization(authorizationSupplier, authorizationType);
+        return this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on encoded authorization token and
+     * {@code Basic} authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param authorizationSupplier A supplier of encoded client authorization token.
+     */
+    protected AbstractOAuth2AuthorizationGrantBuilder clientBasicAuthorization(
+            Supplier<String> authorizationSupplier) {
+        clientAuthorization = ClientAuthorization.ofBasicAuthorization(authorizationSupplier);
+        return this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on client credentials and
+     * authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     * @param authorizationType One of the registered HTTP authentication schemes as per
+     *                          <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">
+     *                          HTTP Authentication Scheme Registry</a>.
+     */
+    protected AbstractOAuth2AuthorizationGrantBuilder clientCredentials(
+            Supplier<Map.Entry<String, String>> credentialsSupplier, String authorizationType) {
+        clientAuthorization = ClientAuthorization.ofCredentials(credentialsSupplier, authorizationType);
+        return this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on client credentials and
+     * {@code Basic} authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     */
+    protected AbstractOAuth2AuthorizationGrantBuilder clientCredentials(
+            Supplier<Map.Entry<String, String>> credentialsSupplier) {
+        clientAuthorization = ClientAuthorization.ofCredentials(credentialsSupplier);
+        return this;
+    }
+
+    /**
+     * A period when the token should be refreshed proactively prior to its expiry.
+     */
+    protected AbstractOAuth2AuthorizationGrantBuilder refreshBefore(Duration refreshBefore) {
+        this.refreshBefore = requireNonNull(refreshBefore, "refreshBefore");
+        return this;
+    }
+
+    @Nullable
+    protected Duration refreshBefore() {
```

Comment:
Could return non-null once we move the default value into this class.

Reply:
fixed